src/passes/referencedLibraries.ts

```diff
@@ -37,46 +31,64 @@ export class ReferencedLibraries extends ASTMapper {
         return this.visitExpression(node, ast);
       }
 
-      //Checks if the Function is a referenced Library functions,
-      //if yes add it to the linearizedBaseContract list of parent ContractDefinition node
-      //free functions calling library functions are not yet supported
-      librariesById.forEach((library, _) => {
-        if (library.vFunctions.some((libraryFunc) => libraryFunc.id === calledDeclaration.id)) {
-          const parent = node.getClosestParentByType(ContractDefinition);
-          if (parent === undefined) return;
+      // Free functions calling library functions are not yet supported
+      const parent = node.getClosestParentByType(ContractDefinition);
+      if (parent === undefined) return;
 
-          getLibrariesToInherit(library, librariesById).forEach((id) => {
-            if (!parent.linearizedBaseContracts.includes(id)) {
-              parent.linearizedBaseContracts.push(id);
-            }
-          });
-        }
-      });
+      // Checks if the Function is a referenced Library function,
+      // if yes add it to the linearizedBaseContract list of parent ContractDefinition node.
+      const library = findFunctionInLibrary(calledDeclaration.id, librariesById);
+      if (library !== undefined) {
+        getLibrariesToInherit(library, librariesById).forEach((id) => {
+          if (!parent.linearizedBaseContracts.includes(id)) {
+            parent.linearizedBaseContracts.push(id);
+          }
+        });
+      }
     }
     this.commonVisit(node, ast);
   }
 }
 
 function getLibrariesToInherit(
   calledLibrary: ContractDefinition,
-  librariesById: Map<number, ASTNode>,
+  librariesById: Map<number, ContractDefinition>,
 ): number[] {
-  const ids: number[] = [calledLibrary.id];
+  const ids: number[] = [];
+  getAllLibraries(calledLibrary, librariesById, ids);
+
+  return ids;
+}
+
+function findFunctionInLibrary(
+  functionId: number,
+  librariesById: Map<number, ContractDefinition>,
+): ContractDefinition | undefined {
+  for (const library of librariesById.values()) {
+    if (library.vFunctions.some((f) => f.id == functionId)) return library;
+  }
+  return undefined;
+}
+
+function getAllLibraries(
+  calledLibrary: ContractDefinition,
+  librariesById: Map<number, ContractDefinition>,
+  ids: number[],
+): void {
+  ids.push(calledLibrary.id);
 
   calledLibrary
     .getChildren()
```

Comment:
Considering every children of the library, the list ids would also have id's of unecessary libraries, Shouldn't we consider the function call which is inside the body of refrenced function ?

Reply:
Yes you are right, but I don't think we need it to get this working @GuyMcComb can you add this as a ticket to optimise and land the pr as is?